JDBCRepositorySQLLoader  native-image之后  loadFromDirectory方法,虽然进行了文件系统的创建,如果存在抛出java.nio.file.FileSystemAlreadyExistsException异常,没有考虑

sharding版本就是当前master的代码
错误堆栈照片:
![4441709173947_ pic](https://github.com/apache/shardingsphere/assets/54206202/9025cd2c-e559-4d7c-aa91-47cf4abf3bab)
错误堆栈:
Caused by: java.nio.file.FileSystemAlreadyExistsException: null
 at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystemProvider.newFileSystem(NativeImageResourceFileSystemProvider.java:115) ~[ai:na]
 at java.base@21.0.2/java.nio.file.FileSystems.newFileSystem(FileSystems.java:339) ~[na:na]
 at java.base@21.0.2/java.nio.file.FileSystems.newFileSystem(FileSystems.java:288) ~[na:na]
 at org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader.loadFromDirectory(JDBCRepositorySQLLoader.java:116) ~[na:na]
 at org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader.load(JDBCRepositorySQLLoader.java:81) ~[na:na]
 at org.apache.shardingsphere.mode.repository.standalone.jdbc.JDBCRepository.init(JDBCRepository.java:59) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.findDefaultService(TypedSPILoader.java:73) ~[na:na]
 at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.findService(TypedSPILoader.java:57) ~[na:na]
 at org.apache.shardingsphere.infra.spi.type.typed.TypedSPILoader.getService(TypedSPILoader.java:110) ~[na:na]
 at org.apache.shardingsphere.mode.manager.standalone.NewStandaloneContextManagerBuilder.build(NewStandaloneContextManagerBuilder.java:49) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.createContextManager(ShardingSphereDataSource.java:78) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.datasource.ShardingSphereDataSource.<init>(ShardingSphereDataSource.java:66) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.api.ShardingSphereDataSourceFactory.createDataSource(ShardingSphereDataSourceFactory.java:95) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.api.yaml.YamlShardingSphereDataSourceFactory.createDataSource(YamlShardingSphereDataSourceFactory.java:135) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.api.yaml.YamlShardingSphereDataSourceFactory.createDataSource(YamlShardingSphereDataSourceFactory.java:70) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.createDataSource(DriverDataSourceCache.java:55) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.lambda$get$0(DriverDataSourceCache.java:48) ~[ai:5.4.2-SNAPSHOT]
 at java.base@21.0.2/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708) ~[ai:na]
 at org.apache.shardingsphere.driver.jdbc.core.driver.DriverDataSourceCache.get(DriverDataSourceCache.java:48) ~[ai:5.4.2-SNAPSHOT]
 at org.apache.shardingsphere.driver.ShardingSphereDriver.connect(ShardingSphereDriver.java:53) ~[ai:5.4.2-SNAPSHOT]
 at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:121) ~[ai:na]
 at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359) ~[ai:na]
 at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201) ~[ai:na]
 at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470) ~[ai:na]
 at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561) ~[ai:na]
 ... 103 common frames omitted


这是我自己clone以下master的代码,进行的判断,它是成功的
![4471709174079_ pic](https://github.com/apache/shardingsphere/assets/54206202/0ae01b00-0e24-4241-aa1a-1da6628e56e7)
成功的打印信息:
2024-02-29T10:56:24.891+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2024-02-29T10:56:24.908+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-2 - Starting...
2024-02-29T10:56:24.914+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.pool.HikariPool        : HikariPool-2 - Added connection conn0: url=jdbc:h2:mem:config user=SA
2024-02-29T10:56:24.914+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-2 - Start completed.
2024-02-29T10:56:24.920+08:00  INFO 8559 --- [               ] com.alibaba.druid.pool.DruidDataSource   : {dataSource-1} inited
2024-02-29T10:56:33.231+08:00  INFO 8559 --- [               ] com.alibaba.druid.pool.DruidDataSource   : {dataSource-2} inited
2024-02-29T10:56:39.004+08:00  INFO 8559 --- [               ] com.alibaba.druid.pool.DruidDataSource   : {dataSource-3} inited
2024-02-29T10:56:41.545+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Driver does not support get/set network timeout for connections. (getNetworkTimeout)
2024-02-29T10:56:41.545+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Added connection org.apache.shardingsphere.driver.jdbc.core.connection.ShardingSphereConnection@59ae1330
2024-02-29T10:56:41.545+08:00  INFO 8559 --- [               ] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
2024-02-29T10:56:41.545+08:00 DEBUG 8559 --- [               ] c.h.m.product.ProductMapper.getList      : ==>  Preparing: select id,`number`,price,original_price,`desc`,create_time from product
2024-02-29T10:56:41.591+08:00 DEBUG 8559 --- [               ] c.h.m.product.ProductMapper.getList      : ==> Parameters: 
2024-02-29T10:56:41.592+08:00  INFO 8559 --- [               ] ShardingSphere-SQL                       : Logic SQL: select id,`number`,price,original_price,`desc`,create_time from product
2024-02-29T10:56:41.592+08:00  INFO 8559 --- [               ] ShardingSphere-SQL                       : Actual SQL: ai-slave1 ::: select id,`number`,price,original_price,`desc`,create_time from product
2024-02-29T10:56:41.650+08:00 DEBUG 8559 --- [               ] c.h.m.product.ProductMapper.getList      : <==      Total: 3

--- a/infra/reachability-metadata/src/main/resources/META-INF/native-image/org.apache.shardingsphere/shardingsphere-infra-reachability-metadata/reflect-config.json
+++ b/infra/reachability-metadata/src/main/resources/META-INF/native-image/org.apache.shardingsphere/shardingsphere-infra-reachability-metadata/reflect-config.json
@@ -189,6 +189,46 @@
   "name":"org.apache.shardingsphere.infra.algorithm.core.yaml.YamlAlgorithmConfigurationCustomizer"
 },
 {
+  "condition":{"typeReachable":"org.apache.shardingsphere.sharding.rule.ShardingRule"},
+  "name":"org.apache.shardingsphere.infra.algorithm.keygen.snowflake.SnowflakeKeyGenerateAlgorithm",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
+  "condition":{"typeReachable":"org.apache.shardingsphere.sharding.rule.ShardingRule"},
+  "name":"org.apache.shardingsphere.infra.algorithm.keygen.uuid.UUIDKeyGenerateAlgorithm",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
+  "condition":{"typeReachable":"org.apache.shardingsphere.readwritesplitting.rule.ReadwriteSplittingRule"},
+  "name":"org.apache.shardingsphere.infra.algorithm.load.balancer.random.RandomLoadBalanceAlgorithm",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
+  "condition":{"typeReachable":"org.apache.shardingsphere.readwritesplitting.rule.ReadwriteSplittingRule"},
+  "name":"org.apache.shardingsphere.infra.algorithm.load.balancer.round.robin.RoundRobinLoadBalanceAlgorithm",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
+  "condition":{"typeReachable":"org.apache.shardingsphere.readwritesplitting.rule.ReadwriteSplittingRule"},
+  "name":"org.apache.shardingsphere.infra.algorithm.load.balancer.weight.WeightLoadBalanceAlgorithm",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
+  "condition":{"typeReachable":"org.apache.shardingsphere.infra.expr.core.InlineExpressionParserFactory"},
+  "name":"org.apache.shardingsphere.infra.expr.groovy.GroovyInlineExpressionParser",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
+  "condition":{"typeReachable":"org.apache.shardingsphere.infra.expr.core.InlineExpressionParserFactory"},
+  "name":"org.apache.shardingsphere.infra.expr.interval.IntervalInlineExpressionParser",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
+  "condition":{"typeReachable":"org.apache.shardingsphere.infra.expr.core.InlineExpressionParserFactory"},
+  "name":"org.apache.shardingsphere.infra.expr.literal.LiteralInlineExpressionParser",
+  "methods":[{"name":"<init>","parameterTypes":[] }]
+},
+{
   "condition":{"typeReachable":"org.apache.shardingsphere.mode.manager.cluster.coordinator.registry.status.compute.service.ComputeNodeStatusService"},
   "name":"org.apache.shardingsphere.infra.instance.ComputeNodeData",
   "allDeclaredFields":true,
@@ -202,16 +242,6 @@
 {
   "condition":{"typeReachable":"org.apache.shardingsphere.mode.manager.cluster.coordinator.registry.status.compute.service.ComputeNodeStatusService"},
   "name":"org.apache.shardingsphere.infra.instance.ComputeNodeDataCustomizer"
-},
-{
-  "condition":{"typeReachable":"org.apache.shardingsphere.sharding.rule.ShardingRule"},
-  "name":"org.apache.shardingsphere.infra.keygen.snowflake.algorithm.SnowflakeKeyGenerateAlgorithm",
-  "methods":[{"name":"<init>","parameterTypes":[] }]
-},
-{
-  "condition":{"typeReachable":"org.apache.shardingsphere.sharding.rule.ShardingRule"},
-  "name":"org.apache.shardingsphere.infra.keygen.uuid.algorithm.UUIDKeyGenerateAlgorithm",
-  "methods":[{"name":"<init>","parameterTypes":[] }]
 },
 {
   "condition":{"typeReachable":"org.apache.shardingsphere.infra.parser.cache.SQLStatementCacheBuilder"},
@@ -545,7 +575,7 @@
   "methods":[{"name":"<init>","parameterTypes":[] }]
 },
 {
-  "condition":{"typeReachable":"org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader$2"},
+  "condition":{"typeReachable":"org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader"},
   "name":"org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQL",
   "allDeclaredFields":true,
   "queryAllDeclaredMethods":true,
@@ -596,21 +626,6 @@
 {
   "condition":{"typeReachable":"org.apache.shardingsphere.infra.util.yaml.YamlEngine"},
   "name":"org.apache.shardingsphere.parser.yaml.config.YamlSQLParserRuleConfigurationCustomizer"
-},
-{
-  "condition":{"typeReachable":"org.apache.shardingsphere.readwritesplitting.rule.ReadwriteSplittingRule"},
-  "name":"org.apache.shardingsphere.readwritesplitting.algorithm.loadbalance.RandomReadQueryLoadBalanceAlgorithm",
-  "methods":[{"name":"<init>","parameterTypes":[] }]
-},
-{
-  "condition":{"typeReachable":"org.apache.shardingsphere.readwritesplitting.rule.ReadwriteSplittingRule"},
-  "name":"org.apache.shardingsphere.readwritesplitting.algorithm.loadbalance.RoundRobinReadQueryLoadBalanceAlgorithm",
-  "methods":[{"name":"<init>","parameterTypes":[] }]
-},
-{
-  "condition":{"typeReachable":"org.apache.shardingsphere.readwritesplitting.rule.ReadwriteSplittingRule"},
-  "name":"org.apache.shardingsphere.readwritesplitting.algorithm.loadbalance.WeightReadQueryLoadBalanceAlgorithm",
-  "methods":[{"name":"<init>","parameterTypes":[] }]
 },
 {
   "condition":{"typeReachable":"org.apache.shardingsphere.infra.util.yaml.YamlEngine"},

--- a/infra/reachability-metadata/src/main/resources/META-INF/native-image/org.apache.shardingsphere/shardingsphere-infra-reachability-metadata/resource-config.json
+++ b/infra/reachability-metadata/src/main/resources/META-INF/native-image/org.apache.shardingsphere/shardingsphere-infra-reachability-metadata/resource-config.json
@@ -13,7 +13,7 @@
     "condition":{"typeReachable":"org.apache.shardingsphere.infra.state.datasource.DataSourceStateManager"},
     "pattern":"\\QMETA-INF/services/java.sql.Driver\\E"
   }, {
-    "condition":{"typeReachable":"org.apache.shardingsphere.mode.repository.standalone.jdbc.sql.JDBCRepositorySQLLoader"},
+    "condition":{"typeReachable":"org.apache.shardingsphere.infra.util.directory.ClasspathResourceDirectoryReader"},
     "pattern":"\\Qsql\\E"
   }, {
     "condition":{"typeReachable":"org.apache.shardingsphere.transaction.xa.atomikos.manager.AtomikosTransactionManagerProvider"},

--- a/infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java
+++ b/infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java
@@ -37,10 +37,8 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.Enumeration;
 import java.util.HashSet;
-import java.util.List;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
@@ -122,7 +120,7 @@ public class ClasspathResourceDirectoryReader {
             if (JAR_URL_PROTOCOLS.contains(directoryUrl.getProtocol())) {
                 return readDirectoryInJar(directory, directoryUrl);
             } else {
-                return readDirectoryInFileSystem(directoryUrl).stream();
+                return readDirectoryInFileSystem(directoryUrl);
             }
         });
     }
@@ -160,31 +158,35 @@ public class ClasspathResourceDirectoryReader {
      * so ShardingSphere need to manually open and close the FileSystem corresponding to the `resource:/` scheme.
      * For more background reference <a href="https://github.com/oracle/graal/issues/7682">oracle/graal#7682</a>.
      * Under the context of third-party dependencies such as Spring Framework OSS,
-     * `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem` will be automatically created during the life cycle of the Context,
+     * `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem` will be automatically created during the life cycle of the context,
      * so additional determination is required.
      *
      * @param directoryUrl directory url
-     * @return list of resource name
+     * @return stream of resource name
      */
     @SneakyThrows({IOException.class, URISyntaxException.class})
-    private static List<String> readDirectoryInFileSystem(final URL directoryUrl) {
+    private static Stream<String> readDirectoryInFileSystem(final URL directoryUrl) {
         if ("resource".equals(directoryUrl.getProtocol())) {
             try (FileSystem ignored = FileSystems.getFileSystem(directoryUrl.toURI())) {
                 return loadFromDirectory(directoryUrl);
             } catch (FileSystemNotFoundException exception) {
-                try (FileSystem ignored = FileSystems.newFileSystem(directoryUrl.toURI(), Collections.emptyMap())) {
-                    return loadFromDirectory(directoryUrl);
-                }
+                FileSystem nativeImageResourceFileSystem = FileSystems.newFileSystem(directoryUrl.toURI(), Collections.emptyMap());
+                return loadFromDirectory(directoryUrl).onClose(() -> {
+                    try {
+                        nativeImageResourceFileSystem.close();
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
             }
         }
         return loadFromDirectory(directoryUrl);
     }
     
-    private static List<String> loadFromDirectory(final URL directoryUrl) throws URISyntaxException, IOException {
+    private static Stream<String> loadFromDirectory(final URL directoryUrl) throws URISyntaxException, IOException {
         Path directoryPath = Paths.get(directoryUrl.toURI());
         // noinspection resource
         Stream<Path> walkStream = Files.find(directoryPath, Integer.MAX_VALUE, (path, basicFileAttributes) -> !basicFileAttributes.isDirectory(), FileVisitOption.FOLLOW_LINKS);
-        return walkStream.map(path -> path.subpath(directoryPath.getNameCount() - 1, path.getNameCount()).toString())
-                .collect(Collectors.toList());
+        return walkStream.map(path -> path.subpath(directoryPath.getNameCount() - 1, path.getNameCount()).toString());
     }
 }

--- a/infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java
+++ b/infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java
@@ -167,7 +167,7 @@ public class ClasspathResourceDirectoryReader {
     @SneakyThrows({IOException.class, URISyntaxException.class})
     private static Stream<String> readDirectoryInFileSystem(final URL directoryUrl) {
         if ("resource".equals(directoryUrl.getProtocol())) {
-            try (FileSystem ignored = FileSystems.getFileSystem(directoryUrl.toURI())) {
+            try {
                 return loadFromDirectory(directoryUrl);
             } catch (FileSystemNotFoundException exception) {
                 FileSystem nativeImageResourceFileSystem = FileSystems.newFileSystem(directoryUrl.toURI(), Collections.emptyMap());

--- a/infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java
+++ b/infra/util/src/main/java/org/apache/shardingsphere/infra/util/directory/ClasspathResourceDirectoryReader.java
@@ -22,11 +22,11 @@ import lombok.SneakyThrows;
 
 import java.io.IOException;
 import java.net.JarURLConnection;
-import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLConnection;
 import java.nio.file.FileSystem;
+import java.nio.file.FileSystemNotFoundException;
 import java.nio.file.FileSystems;
 import java.nio.file.FileVisitOption;
 import java.nio.file.Files;
@@ -37,8 +37,10 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.Enumeration;
 import java.util.HashSet;
+import java.util.List;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
@@ -66,7 +68,7 @@ public class ClasspathResourceDirectoryReader {
      * @param name resource name
      * @return true if the resource is a directory; false if the resource does not exist, is not a directory, or it cannot be determined if the resource is a directory or not.
      */
-    @SneakyThrows({IOException.class, URISyntaxException.class})
+    @SneakyThrows(URISyntaxException.class)
     public static boolean isDirectory(final ClassLoader classLoader, final String name) {
         URL resourceUrl = classLoader.getResource(name);
         if (null == resourceUrl) {
@@ -79,11 +81,6 @@ public class ClasspathResourceDirectoryReader {
             }
             return jarFile.getJarEntry(name).isDirectory();
         } else {
-            if ("resourceUrl".equals(resourceUrl.getProtocol())) {
-                try (FileSystem ignored = FileSystems.newFileSystem(URI.create("resource:/"), Collections.emptyMap())) {
-                    return Files.isDirectory(Paths.get(resourceUrl.toURI()));
-                }
-            }
             return Files.isDirectory(Paths.get(resourceUrl.toURI()));
         }
     }
@@ -125,7 +122,7 @@ public class ClasspathResourceDirectoryReader {
             if (JAR_URL_PROTOCOLS.contains(directoryUrl.getProtocol())) {
                 return readDirectoryInJar(directory, directoryUrl);
             } else {
-                return readDirectoryInFileSystem(directoryUrl);
+                return readDirectoryInFileSystem(directoryUrl).stream();
             }
         });
     }
@@ -162,24 +159,32 @@ public class ClasspathResourceDirectoryReader {
      * This is mainly to align the behavior of `jdk.nio.zipfs.ZipFileSystem`,
      * so ShardingSphere need to manually open and close the FileSystem corresponding to the `resource:/` scheme.
      * For more background reference <a href="https://github.com/oracle/graal/issues/7682">oracle/graal#7682</a>.
+     * Under the context of third-party dependencies such as Spring Framework OSS,
+     * `com.oracle.svm.core.jdk.resources.NativeImageResourceFileSystem` will be automatically created during the life cycle of the Context,
+     * so additional determination is required.
      *
      * @param directoryUrl directory url
-     * @return stream of resource name
+     * @return list of resource name
      */
     @SneakyThrows({IOException.class, URISyntaxException.class})
-    private static Stream<String> readDirectoryInFileSystem(final URL directoryUrl) {
+    private static List<String> readDirectoryInFileSystem(final URL directoryUrl) {
         if ("resource".equals(directoryUrl.getProtocol())) {
-            try (FileSystem ignored = FileSystems.newFileSystem(URI.create("resource:/"), Collections.emptyMap())) {
+            try (FileSystem ignored = FileSystems.getFileSystem(directoryUrl.toURI())) {
                 return loadFromDirectory(directoryUrl);
+            } catch (FileSystemNotFoundException exception) {
+                try (FileSystem ignored = FileSystems.newFileSystem(directoryUrl.toURI(), Collections.emptyMap())) {
+                    return loadFromDirectory(directoryUrl);
+                }
             }
         }
         return loadFromDirectory(directoryUrl);
     }
     
-    private static Stream<String> loadFromDirectory(final URL directoryUrl) throws URISyntaxException, IOException {
+    private static List<String> loadFromDirectory(final URL directoryUrl) throws URISyntaxException, IOException {
         Path directoryPath = Paths.get(directoryUrl.toURI());
         // noinspection resource
         Stream<Path> walkStream = Files.find(directoryPath, Integer.MAX_VALUE, (path, basicFileAttributes) -> !basicFileAttributes.isDirectory(), FileVisitOption.FOLLOW_LINKS);
-        return walkStream.map(path -> path.subpath(directoryPath.getNameCount() - 1, path.getNameCount()).toString());
+        return walkStream.map(path -> path.subpath(directoryPath.getNameCount() - 1, path.getNameCount()).toString())
+                .collect(Collectors.toList());
     }
 }

